plugins/woocommerce-blocks/assets/js/atomic/utils/register-block-single-product-template.ts

```diff
@@ -1,140 +1,304 @@
+			}
+		} catch ( error ) {
+			// eslint-disable-next-line no-console
+			console.debug(
+				`Failed to unregister block ${ blockName }:`,
+				error
 			);
-			const hasChangedTemplate = previousTemplateId !== currentTemplateId;
+		}
+	}
+
+	/**
+	 * Registers a block or block variation.
+	 * Handles different registration requirements for various contexts.
+	 * Includes checks to prevent recursive registration.
+	 *
+	 * @param {BlockConfig} config - Configuration of the block to register
+	 */
+	private registerBlock( config: BlockConfig ): void {
+		const {
+			blockName,
+			blockMetadata = blockName,
+			blockSettings,
+			isVariationBlock,
+			variationName,
+			isAvailableOnPostEditor,
+		} = config;
 
-			if ( ! hasChangedTemplate || ! blockName ) {
+		try {
+			// Check if block is already registered
+			const key = variationName || blockName;
+			if ( this.hasAttemptedRegistration( key ) ) {
 				return;
 			}
 
-			let isRegistered = isBlockRegistered( blockName );
-
-			/**
-			 * We need to unregister the block each time the user visits or leaves the Single Product template.
-			 *
-			 * The Single Product template is the only template where the `ancestor` property is not needed because it provides the context
-			 * for the product blocks. We need to unregister and re-register the block to remove or add the `ancestor` property depending on which
-			 * location (template, post, page, etc.) the user is in.
-			 *
-			 */
-			if (
-				isRegistered &&
-				( currentTemplateId?.includes( 'single-product' ) ||
-					previousTemplateId?.includes( 'single-product' ) )
-			) {
-				if ( isVariationBlock && variationName ) {
-					unregisterBlockVariation( blockName, variationName );
-				} else {
-					unregisterBlockType( blockName );
-				}
-				isRegistered = false;
+			const editSiteStore = select( 'core/edit-site' );
+
+			// Don't register if we're in post editor context and block isn't available there
+			if ( ! editSiteStore && ! isAvailableOnPostEditor ) {
+				return;
 			}
 
-			if ( ! isRegistered ) {
-				if ( isVariationBlock ) {
-					// @ts-expect-error: `registerBlockType` is not typed in WordPress core
-					registerBlockVariation( blockName, blockSettings );
-				} else {
-					const ancestor = isEmpty( blockSettings?.ancestor )
-						? [ 'woocommerce/single-product' ]
-						: blockSettings?.ancestor;
-					// @ts-expect-error: `registerBlockType` is not typed in WordPress core
-					registerBlockType( blockMetadata, {
-						...blockSettings,
-						ancestor: ! currentTemplateId?.includes(
-							'single-product'
-						)
-							? ancestor
-							: undefined,
-					} );
-				}
+			if ( isVariationBlock ) {
+				registerBlockVariation(
+					blockName,
+					blockSettings as BlockVariation< BlockAttributes >
+				);
+			} else {
+				const ancestor = isEmpty( blockSettings?.ancestor )
+					? [ 'woocommerce/single-product' ]
+					: blockSettings?.ancestor;
```

Comment:
This is another logic specific for the blocks related to the Single Product Template/Single Product Block. I'm fine to keep it as it is.